adrs/2023-02-10-limit-manager-role-permission.md

```diff
@@ -0,0 +1,79 @@
+
+- Get/List/Create/Delete/Update/Patch/Watch on `AutoScalingRunnerSets` (with `Status` and `Finalizer` sub-resource)
+- Get/List/Create/Delete/Update/Patch/Watch on `AutoScalingListeners` (with `Status` and `Finalizer` sub-resource)
+- Get/List/Create/Delete/Update/Patch/Watch on `EphemeralRunnerSets` (with `Status` and `Finalizer` sub-resource)
+- Get/List/Create/Delete/Update/Patch/Watch on `EphemeralRunners` (with `Status` and `Finalizer` sub-resource)
+
+- Get/List/Create/Delete/Update/Patch/Watch on `Pods` (with `Status` sub-resource)
+- **Get/List/Create/Delete/Update/Patch/Watch on `Secrets`**
+- Get/List/Create/Delete/Update/Patch/Watch on `Roles`
+- Get/List/Create/Delete/Update/Patch/Watch on `RoleBindings`
+- Get/List/Create/Delete/Update/Patch/Watch on `ServiceAccounts`
+
+> Full list can be found at: https://github.com/actions/actions-runner-controller/blob/facae69e0b189d3b5dd659f36df8a829516d2896/charts/actions-runner-controller-2/templates/manager_role.yaml
+
+## Limit cluster role permission on Secrets
+
+The cluster scope `List` `Secrets` permission might be a blocker for adopting `actions-runner-controller` for certain customers as they may have certain restriction in their cluster that simply doesn't allow any service account to have cluster scope `List Secrets` permission. 
+
+To help these customers and improve security for `actions-runner-controller` in general, we will try to limit the `ClusterRole` permission of the controller manager's service account down to the following:
+
+- Get/List/Create/Delete/Update/Patch/Watch on `AutoScalingRunnerSets` (with `Status` and `Finalizer` sub-resource)
+- Get/List/Create/Delete/Update/Patch/Watch on `AutoScalingListeners` (with `Status` and `Finalizer` sub-resource)
+- Get/List/Create/Delete/Update/Patch/Watch on `EphemeralRunnerSets` (with `Status` and `Finalizer` sub-resource)
+- Get/List/Create/Delete/Update/Patch/Watch on `EphemeralRunners` (with `Status` and `Finalizer` sub-resource)
+
+- List/Watch on `Pods`
+- List/Watch on `Roles`
+- List/Watch on `RoleBindings`
+- List/Watch on `ServiceAccounts`
+
+> We will change the default cache-based client to bypass cache on reading `Secrets`, so we can eliminate the need for `List` and `Watch` `Secrets` permission in cluster scope.
+
+Introduce a new `ClusterRole` for the namespace that each `AutoScalingRunnerSet` deployed with the following permission
+
+- Get/Create/Delete/Update/Patch on `Secrets`
+- Get/Create/Delete/Update/Patch on `Pods`
+- Get/Create/Delete/Update/Patch on `Roles`
+- Get/Create/Delete/Update/Patch on `RoleBindings`
+- Get/Create/Delete/Update/Patch on `ServiceAccounts`
+
+The `RoleBinding` for the new `ClusterRole` will happen during `helm install demo oci://ghcr.io/actions/actions-runner-controller-charts/auto-scaling-runner-set` to grant the controller's service account required permissions to operate in the namespace the `AutoScalingRunnerSet` deployed.
```

Comment:
If we ensure that the `RoleBinding` is used here by default, then it's definitely acceptable from our side. However, I'm wondering why do we favour `ClusterRole` here if it's never going to be granted on the cluster level. Is it done just for the sake of simplicity i.e. to avoid explicitly specifying the namespace in the role definition? If that's the case it's fine to me.

Besides, I'm not sure if the statement: "`ClusterRole` for the namespace" is correct. ClusterRole is not really scoped to any namespace per its definition. It can only be scoped during the binding.

See:
https://kubernetes.io/docs/reference/access-authn-authz/rbac/#clusterrole-example

Reply:
The idea is to have a single `ClusterRole` and multiple `RoleBinding` to bind the cluster role into each namespace.
Otherwise, we will need to create a `Role` in each namespace.
Again, sorry for the confusion... I will keep updating the doc.

Reply:
⭐ I presume ARC should have two types of installation, cluster-wide and namespace-isolated.
Using ClusterRole universally and combining it with either RoleBinding or ClusterRoleBinding depending on the desired config would enable users to choose the right balance between security and operability while keeping the chart DRY.

Reply:
@mumoshu I will add a new section in the doc to talk about that. 🙇